Fix GCloud IT: `test_max_messages` error (1/2)

### Description

GCloud IT suite fails in the [nightly](https://github.com/wazuh/wazuh-qa/issues/2906) pipeline.
Is required further research to determine the cause of the error.



### Failed cases
- `test_gcloud/test_functionality/test_max_messages.py::test_max_messages[get_configuration0-publish_messages1]`
- `test_gcloud/test_functionality/test_max_messages.py::test_max_messages[get_configuration0-publish_messages2]`

<details>
test_max_messages[get_configuration0-publish_messages1]:

```
test_gcloud/test_functionality/test_max_messages.py:183: in test_max_messages
    assert int(number_pulled) >= publish_messages
E   AssertionError: assert 0 >= 100
E    +  where 0 = int('0')
```

test_max_messages[get_configuration0-publish_messages2]

```
test_gcloud/test_functionality/test_max_messages.py:199: in test_max_messages
    assert int(number_pulled) >= remainder
E   AssertionError: assert 0 >= 20
E    +  where 0 = int('0')
```

</details>

### Build
https://ci.wazuh.info/job/Test_integration/26273/

### Report 
 [report.html](https://github.com/wazuh/wazuh-qa/files/8677359/manager_html_report_Test_integration_B26273_20220512005724.1.zip)

---------
- [x]  analyze test_max_messages[get_configuration0-publish_messages1]
- [x]  analyze test_max_messages[get_configuration0-publish_messages2]

--- a/deps/wazuh_testing/wazuh_testing/gcloud.py
+++ b/deps/wazuh_testing/wazuh_testing/gcloud.py
@@ -56,9 +56,10 @@ def detect_gcp_start(file_monitor):
 
 
 def callback_received_messages_number(line):
-    match = re.match(r'.*wm_gcp_pubsub_run\(\): INFO: - INFO - Received and acknowledged (\d+) messages', line)
+    received_messages_number = rf".*wm_gcp_pubsub_run\(\): INFO: - INFO - Received and acknowledged (\d+) messages"
+    match = re.findall(received_messages_number, line)
     if match:
-        return match.group(1)
+        return match
     return None
 
 

--- a/tests/integration/test_gcloud/test_functionality/test_max_messages.py
+++ b/tests/integration/test_gcloud/test_functionality/test_max_messages.py
@@ -59,6 +59,7 @@ tags:
     - scan
     - maximum
 '''
+from itertools import count
 import os
 import sys
 
@@ -82,6 +83,7 @@ interval = '25s'
 pull_messages_timeout = global_parameters.default_timeout + 60
 pull_on_start = 'no'
 max_messages = 100
+count_message = 0
 logging = 'info'
 wazuh_log_monitor = FileMonitor(LOG_FILE_PATH)
 test_data_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'data')
@@ -190,8 +192,10 @@ def test_max_messages(get_configuration, configure_environment, reset_ossec_log,
         # GCP might log messages from sources other than ourselves
         for number_pulled in numbers_pulled:
             if int(number_pulled) != 0:
-                assert int(number_pulled) >= publish_messages
+                if (int(number_pulled) >= publish_messages):
+                    count_message += 1
                 assert int(number_pulled) <= max_messages
+        assert count_message >=1
     else:
         for number_pulled in numbers_pulled:
             if int(number_pulled) != 0:

--- a/tests/integration/test_gcloud/test_functionality/test_max_messages.py
+++ b/tests/integration/test_gcloud/test_functionality/test_max_messages.py
@@ -124,7 +124,8 @@ def get_configuration(request):
     ['- DEBUG - GCP message' for _ in range(100)],
     ['- DEBUG - GCP message' for _ in range(120)]
 ], indirect=True)
-def test_max_messages(get_configuration, configure_environment, reset_ossec_log, publish_messages, daemons_handler, wait_for_gcp_start):
+def test_max_messages(get_configuration, configure_environment, reset_ossec_log, publish_messages,
+                      daemons_handler, wait_for_gcp_start):
     '''
     description: Check if the 'gcp-pubsub' module pulls a message number less than or equal to the limit set
                  in the 'max_messages' tag. For this purpose, the test will use a fixed limit and generate a
@@ -195,7 +196,7 @@ def test_max_messages(get_configuration, configure_environment, reset_ossec_log,
                     # A counter is used to prevent it from failing due to logs from other sources
                     count_message += 1
                 assert int(number_pulled) <= max_messages
-        assert count_message >=1
+        assert count_message >= 1
     else:
         for number_pulled in numbers_pulled:
             if int(number_pulled) != 0:

--- a/tests/integration/test_gcloud/test_functionality/test_max_messages.py
+++ b/tests/integration/test_gcloud/test_functionality/test_max_messages.py
@@ -59,11 +59,10 @@ tags:
     - scan
     - maximum
 '''
-from itertools import count
 import os
 import sys
-
 import pytest
+
 from wazuh_testing import global_parameters
 from wazuh_testing.fim import generate_params
 from wazuh_testing.gcloud import callback_detect_start_fetching_logs, callback_received_messages_number
@@ -193,6 +192,7 @@ def test_max_messages(get_configuration, configure_environment, reset_ossec_log,
         for number_pulled in numbers_pulled:
             if int(number_pulled) != 0:
                 if (int(number_pulled) >= publish_messages):
+                    # A counter is used to prevent it from failing due to logs from other sources
                     count_message += 1
                 assert int(number_pulled) <= max_messages
         assert count_message >=1

--- a/tests/integration/test_gcloud/test_functionality/test_max_messages.py
+++ b/tests/integration/test_gcloud/test_functionality/test_max_messages.py
@@ -199,5 +199,4 @@ def test_max_messages(get_configuration, configure_environment, reset_ossec_log,
         assert count_message >= 1
     else:
         for number_pulled in numbers_pulled:
-            if int(number_pulled) != 0:
-                assert int(number_pulled) <= max_messages
+            assert int(number_pulled) <= max_messages

--- a/tests/integration/test_gcloud/test_functionality/test_max_messages.py
+++ b/tests/integration/test_gcloud/test_functionality/test_max_messages.py
@@ -109,6 +109,7 @@ truncate_file(LOG_FILE_PATH)
 
 # fixtures
 
+
 @pytest.fixture(scope='module', params=configurations)
 def get_configuration(request):
     """Get configurations from the module."""
@@ -180,11 +181,11 @@ def test_max_messages(get_configuration, configure_environment, reset_ossec_log,
                             callback=callback_detect_start_fetching_logs,
                             error_message='Did not receive expected '
                                           '"Starting fetching of logs" event')
-    
+
     numbers_pulled = wazuh_log_monitor.start(timeout=pull_messages_timeout,
-                                                callback=callback_received_messages_number,
-                                                error_message='Did not receive expected '
-                                                              '- INFO - Received and acknowledged x messages').result()
+                                             callback=callback_received_messages_number,
+                                             error_message='Did not receive expected '
+                                                           '- INFO - Received and acknowledged x messages').result()
     if publish_messages <= max_messages:
         # GCP might log messages from sources other than ourselves
         for number_pulled in numbers_pulled:
@@ -194,4 +195,4 @@ def test_max_messages(get_configuration, configure_environment, reset_ossec_log,
     else:
         for number_pulled in numbers_pulled:
             if int(number_pulled) != 0:
-                assert int(number_pulled) <= max_messages 
+                assert int(number_pulled) <= max_messages

--- a/tests/integration/test_gcloud/test_functionality/test_max_messages.py
+++ b/tests/integration/test_gcloud/test_functionality/test_max_messages.py
@@ -83,7 +83,6 @@ interval = '25s'
 pull_messages_timeout = global_parameters.default_timeout + 60
 pull_on_start = 'no'
 max_messages = 100
-count_message = 0
 logging = 'info'
 wazuh_log_monitor = FileMonitor(LOG_FILE_PATH)
 test_data_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'data')
@@ -175,6 +174,7 @@ def test_max_messages(get_configuration, configure_environment, reset_ossec_log,
         - logs
         - scheduled
     '''
+    count_message = 0
     str_interval = get_configuration['sections'][0]['elements'][4]['interval']['value']
     time_interval = int(''.join(filter(str.isdigit, str_interval)))
 

--- a/tests/integration/test_gcloud/test_functionality/test_max_messages.py
+++ b/tests/integration/test_gcloud/test_functionality/test_max_messages.py
@@ -1,5 +1,5 @@
 '''
-copyright: Copyright (C) 2015-2021, Wazuh Inc.
+copyright: Copyright (C) 2015-2022, Wazuh Inc.
 
            Created by Wazuh, Inc. <info@wazuh.com>.
 
@@ -180,27 +180,18 @@ def test_max_messages(get_configuration, configure_environment, reset_ossec_log,
                             callback=callback_detect_start_fetching_logs,
                             error_message='Did not receive expected '
                                           '"Starting fetching of logs" event')
-
+    
+    numbers_pulled = wazuh_log_monitor.start(timeout=pull_messages_timeout,
+                                                callback=callback_received_messages_number,
+                                                error_message='Did not receive expected '
+                                                              '- INFO - Received and acknowledged x messages').result()
     if publish_messages <= max_messages:
-        number_pulled = wazuh_log_monitor.start(timeout=pull_messages_timeout,
-                                                callback=callback_received_messages_number,
-                                                error_message='Did not receive expected '
-                                                              '- INFO - Received and acknowledged x messages').result()
         # GCP might log messages from sources other than ourselves
-        assert int(number_pulled) >= publish_messages
+        for number_pulled in numbers_pulled:
+            if int(number_pulled) != 0:
+                assert int(number_pulled) >= publish_messages
+                assert int(number_pulled) <= max_messages
     else:
-        ntimes = int(publish_messages / max_messages)
-        remainder = int(publish_messages % max_messages)
-
-        for i in range(ntimes):
-            number_pulled = wazuh_log_monitor.start(timeout=pull_messages_timeout,
-                                                    callback=callback_received_messages_number,
-                                                    error_message='Did not receive expected '
-                                                                  'Received and acknowledged x messages').result()
-            assert int(number_pulled) == max_messages
-        number_pulled = wazuh_log_monitor.start(timeout=pull_messages_timeout,
-                                                callback=callback_received_messages_number,
-                                                error_message='Did not receive expected '
-                                                              '- INFO - Received and acknowledged x messages').result()
-        # GCP might log messages from sources other than ourselves
-        assert int(number_pulled) >= remainder
+        for number_pulled in numbers_pulled:
+            if int(number_pulled) != 0:
+                assert int(number_pulled) <= max_messages 

--- a/tests/integration/test_gcloud/test_functionality/test_max_messages.py
+++ b/tests/integration/test_gcloud/test_functionality/test_max_messages.py
@@ -188,6 +188,8 @@ def test_max_messages(get_configuration, configure_environment, reset_ossec_log,
                                              callback=callback_received_messages_number,
                                              error_message='Did not receive expected '
                                                            '- INFO - Received and acknowledged x messages').result()
+    # Validate that the number pulled is at least once greater or equal than the published messages
+    # and less than the maximum number of messages extracted allowed in each iteration
     if publish_messages <= max_messages:
         # GCP might log messages from sources other than ourselves
         for number_pulled in numbers_pulled:
@@ -198,5 +200,6 @@ def test_max_messages(get_configuration, configure_environment, reset_ossec_log,
                 assert int(number_pulled) <= max_messages
         assert count_message >= 1
     else:
+        # Validate that the number pulled always is less than the maximum messages pulled allowed in each iteration
         for number_pulled in numbers_pulled:
             assert int(number_pulled) <= max_messages

